src/storage-controller/src/lib.rs

```diff
@@ -1768,27 +1746,42 @@ where
     /// Drops the read capability for the sinks and allows their resources to be reclaimed.
     fn drop_sinks(
         &mut self,
+        storage_metadata: &StorageMetadata,
         identifiers: Vec<GlobalId>,
     ) -> Result<(), StorageError<Self::Timestamp>> {
         self.validate_export_ids(identifiers.iter().cloned())?;
-        self.drop_sinks_unvalidated(identifiers);
+        self.drop_sinks_unvalidated(storage_metadata, identifiers);
         Ok(())
     }
 
-    fn drop_sinks_unvalidated(&mut self, identifiers: Vec<GlobalId>) {
-        for id in identifiers {
-            // Already removed.
-            if self.export(id).is_err() {
-                continue;
-            }
+    fn drop_sinks_unvalidated(
+        &mut self,
+        storage_metadata: &StorageMetadata,
+        mut identifiers: Vec<GlobalId>,
+    ) {
+        // Ignore exports that have already been removed.
+        identifiers.retain(|id| self.export(*id).is_ok());
 
-            // We don't explicitly remove read capabilities! Downgrading the
-            // frontier of the sink to `[]` (the empty Antichain), will
-            // propagate to the storage dependencies.
+        // TODO: ideally we'd advance the write frontier ourselves here, but this function's
```

Comment:
By this you mean the write frontier of the (new) "progress shard"?

Reply:
Yeah, that's right! It's safe to do - the sink code is robust to it - and it would simplify the finalization logic if we could assume that a dropped sink had an empty write / implied frontier.